# Exercise 04: Regression I


Welcome to the fourth exercise for Applied Machine Learning. 

Your objectives for this session are to: 
- learn the basics of the fitting a model to data, 
- fit linear regression and logistic regression models with `sklearn`, and
- interpret the outputs of the models.

____

### Part 1: Fitting a model to data with `sklearn`

Last time we looked into `sklearn` and how to use this python library to implement decision trees. Today we will use the same library to implement linear and logstic regression. 

Linear regression predicts a continuous value for a variable based on the value of another variable. The classic ***regression*** task with a continuous target variable.

Logistic regression predicts a class value for a a variable based on the value of another variable. Although the model has regression in its name, logistic regression is for ***classification*** tasks with a categorical target variable.

Let's import our libraries for today.

In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression

Remember what we learned last week about `sklearn`? We need to train (or "fit") a pre-programmed algorithm (or "estimator") to some data so that it can be used make predictions on new data.

Like last time, the data should be structured as two arrays:

* `X` = 2D array of samples/instances/observations (sometimes called the "samples matrix," "design matrix," or "features matrix"), where instances are represented as rows and features/attributes are represented as columns.

* `y` = 1D array where the ith entry corresponds to the target of the ith sample (row) of X. For regression tasks the target values are real numbers. For classification tasks the target values are integers (or any other discrete set of values). 


Let's look at an example with the same basic structure as last week's example, but this time with just one attribute: 

In [2]:
# define our example model - we will use the Linear Regression model
example_lr = LinearRegression() 

# make some fake samples for X — 4 instances, 1 atrribute
X = [[1],  
     [11],
     [21],
     [31]]

# make some fake target values for y - continuous numeric values because regression!
y = [2.5, 16.2, 24.9, 38.1]  

# fit our example classifier to the data
example_lr.fit(X, y)

# print the score
print("Score: {:.3f}".format(example_lr.score(X, y)))

Score: 0.993


Not so different from last week, right? 

The main change from last week is that today we won't use the `predict` function — we'll just check the `score` to see how well we can fit a model to the data given features and a target. 

However, one other thing we want to do today is inspect the coefficients of our regression models. This is done like so: 

In [3]:
coefficient = example_lr.coef_
print(coefficient)

[1.155]


Remember how to interpret this from lecture? The coefficient is the slope of the regression line. More specifically, this tells value tells us the average change in our dependent variable (`y`) for one-unit change in our independent variable (or attribute, `x`). 

____

### Part 2: Introduction to the dataset

Today we will be working with `HomesSoldHellerup.csv`. A Dataset of housing sales in Hellerup. Each row consists of a home that was sold and the columns represent different attributes of the home. The attributes (columns) are listed for you below:

Attribute Information
- 1) `Road name`: The name of the road
- 2) `Road number`: The road number of the home
- 3) `Type`: What type of home is it? ("Apartment", "Villa" etc.)
- 4) `m2`: The area of the home given in square meters
- 5) `build year`: The year the home was build
- 6) `Zip code`: The zip code which the home lies within
- 7) `City`: The city the home is located in
- 8) `Date of sale`: The day the house was sold
- 9) `Type of sale`: How was the home sold? 
- 10) `Price`: The final price of the home in DKK

Let's see if we can use these attributes to fit a linear regression model capable of estimating the price of a home.

# <font color='red'>TASK 1</font>

It's always a good idea to get an overview of the data we are about to work with. Start by reading the `HomesSoldHellerup.csv` file, using the `read_csv` method from `pandas`, and call the dataset `homes_df`. Then use the `value_counts()` method to check how many of each `Type` of home there is in the dataset. Then inspect the range of `Price`. 


*Hint: Note that the separator in this dataset is a semicolon and not a comma, which is the default. So you need to specify this with a parameter in the `read_csv()` function.*

In [5]:
# your code here
homes_sold = pd.read_csv("HomesSoldHellerup.csv" , sep=";")

### Part 3: Fitting a linear regression model to the data

# <font color='red'>TASK 2</font>

To use a regression model we need at least one attribute and one target varaiable. The target is what we wish to predict based on one or more attributes.  We are going to use Linear Regression using the `sklearn` library and we will therefore use much of the same syntax and functions as we did in week 3. 

Define your feature matrix `X` with the single feature, `m2`, which is the size of the home in square meters. Then define your target `y` as `Price` from the dataset. 

In [12]:
# define X and y
X = homes_sold[["m2"]]
y = homes_sold["Price"]

# <font color='red'>TASK 3</font>
Let's fit a simple linear regression model to `X` and `y`. If you have defined them correctly, this will estimate a coefficient based on the linear relationship between the size of the home in square meters and the price it was sold for. It could also potentially be used to predict the selling price of a home.

Try it. Fit a linear regression model to `X` and `y`, and report the score and the coefficient.


*Hint: Check out the documentation on the `sklearn` site for the linear regression to get a better idea of how it works:* https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html.

In [15]:
# fit a linear regression model to X and y 
reg = LinearRegression().fit(X, y)


In [16]:
# print the score
reg.score(X, y)

0.3048183473178293

What is the "score" for this model? What measure is it exactly?

In [17]:
# print the coefficient
reg.coef_

array([36683.88180408])

What does this coefficient suggest about the relationship between `m2` and `Price`?

### Part 4: Fitting a linear regression model to the data, with a categorical attribute

# <font color='red'>TASK 4</font>
Last week you learned how to make dummy variables from a categorical variable with the `get_dummies` method from the `pandas` library. Dummy variables are especially important for the implementation of regression models, so let's try it here.

Build a linear regression model just like you did in Part 3, but re-define `X` to use only the `Type` of home as the attribute this time. Since this is a categorical attribute, you must first make dummy variables.

Fit the model to `X` and `y`, then report and interpret the score and the coefficient

In [26]:
# re-define X using the `Type` attribute
X = pd.get_dummies(homes_sold[["Type"]], columns=["Type"], drop_first=True)
print(X.head())

   Type_Erhverv  Type_Lejlighed  Type_Rækkehus  Type_Stuehus  Type_Villa
0         False            True          False         False       False
1         False            True          False         False       False
2         False            True          False         False       False
3         False            True          False         False       False
4         False            True          False         False       False


In [27]:
# fit model
reg = LinearRegression().fit(X, y)

In [28]:
# print score
reg.score(X, y)

0.08351211431025096

Based on this score, is this model better or worse than the one you fit with `m2` as the attribute?

Use the code below to extract the regression coefficients, align them with the column names, and then print them.

In [29]:
# Extract the coefficients
coefficients = reg.coef_

# Get the names of the dummy variable columns
column_names = X.columns

# Create a dictionary with column names as keys and coefficients as values
coef_dict = dict(zip(column_names, coefficients))

# Print the dictionary
print(coef_dict)

{'Type_Erhverv': 9252500.000000244, 'Type_Lejlighed': -3082904.4373483406, 'Type_Rækkehus': -1581394.6436780554, 'Type_Stuehus': -3410151.9999999073, 'Type_Villa': -418120.711590196}


How do we interpret this? Why is there not just one coefficient value?

*Hint: Are your coefficients all identical? If so, you've been affected by the ["dummy variable trap"](https://www.statology.org/dummy-variable-trap/). Go back to where you used `get_dummies`, set the argument `drop_first = True`, and re-run your code.*  

*The dummy variable trap generally does not affect a model's predictive performance (i.e., the `score` is not affected). However, when interpreting regression coefficients, we need to avoid multicollinearity — we don't want an attribute to be correlated with another, because then it's hard to distinguish the individual effects of the correlated attributes. If we straightforwardly create a dummy variable for each possible value of a categorical variable, then we often introduce multicollinearity (e.g., if we know that a property is a villa, then we also know that it's not an apartment, business, etc.).*

*The most common way to avoid the dummy variable trap is to create `n-1` dummy variables for a categorical variable with `n` unique values. We do this by setting `drop_first = True` with the `get_dummies` method.*

### Part 5: Fitting a logistic regression model to the data

So far we've been predicting price as a continuous target variable. But let's say, for illustrative purposes, we only care about whether the price is above or below the median price of all the houses on the market. Maybe I want to brag to my friends about having a home that costs more than at least 50% of all homes... I'd be a horrible dinner guest, but this makes for a nice classification task, so let's go with it.

# <font color='red'>TASK 5</font>

Add a new column to your main dataset, `homes_df`, which labels instances where the price is below the median price as 0 (indicating a "low" price) and all other instances 1 (indicating a "high" price). Name the column `Price_binary`.

In [33]:
# your code here
homes_sold["Price_binary"] =  (homes_sold["Price"] >= homes_sold["Price"].median()).astype(int)
homes_sold.tail()

Road name Road Number   Type   m2  Build Year  ZipCode      City  \
2155    Wiehesvej            13  Villa  163        1900     2900  Hellerup   
2156    Wiehesvej            10  Villa  128        1915     2900  Hellerup   
2157  Vinrankevej             9  Villa  182        1963     2900  Hellerup   
2158  Vinrankevej             5  Villa  202        2007     2900  Hellerup   
2159  Vinrankevej             3  Villa  202        2011     2900  Hellerup   

     Date of Sale Type of Sale     Price  Price_binary  
2155     22-06-15    Alm. Salg   8700000             1  
2156     10-06-10    Alm. Salg   9100000             1  
2157     17-09-13    Alm. Salg   7000000             1  
2158     29-08-12    Alm. Salg  13500000             1  
2159     09-07-10    Alm. Salg   4000000             1

# <font color='red'>TASK 6</font>
Now we have a binary, categorical variable that we want to predict: `Price_binary`. So, we no longer have a regression task on our hand, but a classification task. For this we'll use logistic regression.

Implement a logistic regression model that uses `m2` as an attribute and `Price_binary` as a target variable. Fit the model to `X` and `y`, then report the score and the coefficient.

In [34]:
# re-define X and Y
X = homes_sold[["m2"]]
y = homes_sold["Price_binary"]

In [35]:
# fit model
lr = LogisticRegression().fit(X, y)

In [37]:
# print score
lr.score(X, y)

0.8314814814814815

What is the "score" for this logistic regression model? What measure is it exactly?

In [38]:
# print coefficient
lr.coef_

array([[0.03555657]])

How do you interpret this coefficient? What does it tell us? Remember, it's a bit more confusing than linear regression.

So, the above tells us that, on average, for an increase of 1 square meter in size, the odds that the house will have a "high" price is 1.036 — the odds increase a little bit (a 3.6% increase). An odds ratio of 1.00 would mean absolutely no effect. 

---------
**That's it for this week! Next week we'll build regression models with multiple attributes and see how well our models generalize.**